pkg/storage/replica.go

```diff
@@ -1519,6 +1519,18 @@ func (ec *endCmds) doneLocked(
 					// create a transaction record with WriteTooOld set.
 					key := keys.TransactionKey(header.Key, *cr.txnID)
 					cr.txn = roachpb.Span{Key: key}
+				case *roachpb.ScanRequest:
+					resp := br.Responses[i].GetInner().(*roachpb.ScanResponse)
+					if ec.ba.Header.MaxSpanRequestKeys != 0 &&
+						ec.ba.Header.MaxSpanRequestKeys == int64(len(resp.Rows)) {
+						// If the scan requested a limited number of results and we hit the
```

Comment:
Yeah, I'll expand the comment here. We have to add a span that contains all keys that were read, not just the ones that are returned. If we didn't actually stop reading keys, this means we scanned to the end of the range. If we always truncated to the last key returned (i.e. when there is no limit), we wouldn't have proper serializability on non-existent keys that another transaction tries to write. The write to those new keys needs to happen after the read (which is what the timestamp cache enforces).